Fix inconsistent tooltips on system page and menu

martinpitt opened the pull request:
See individual commits for details.

Fixes #8829

martinpitt commented:
See #8829 for screenshots of current master: We use three different tooltip styles on one page!

Now the "hardware info", "machine ID", time, and tuned table cells all use the standard bootstrap notifications, and the blue (i) buttons are gone:

![tooltip-system-fixed](https://user-images.githubusercontent.com/200109/37401928-02ff36d6-278a-11e8-9482-d06141c53b44.png)

![tooltip-machineid](https://user-images.githubusercontent.com/200109/37491269-88691576-289d-11e8-864b-296ae25146c0.png)

![tooltip-ntp](https://user-images.githubusercontent.com/200109/37491274-8b15ac4e-289d-11e8-8f9c-38860f17035e.png)

![tooltip-tuned](https://user-images.githubusercontent.com/200109/37491280-8dcdf0c2-289d-11e8-8fc6-3ca37c40d379.png)

Extra care is taken for the machine ID, this uses a monospaced font now; according to @garrett this is more appropriate, and it also makes maximum width calculation easier (expressed in terms of #digits).

martinpitt commented:
I added another commit to fix #8827

martinpitt commented:
Need to adjust the [tuned test on rhel-7-4](https://fedorapeople.org/groups/cockpit/logs/pull-8835-20180316-012929-b7f1d6db-verify-rhel-7-4/log.html#209) to the old behaviour.

martinpitt commented:
tuned test fix for RHEL 7.4 pushed.

mvollmer commented:
> the blue (i) buttons are gone:

Are the spinner and red (!) button still there?

mvollmer commented on pkg/systemd/host.js:
```diff
@@ -296,20 +292,10 @@ PageServer.prototype = {
-            }
-
-            var icon_html = Mustache.render(self.ntp_status_icon_tmpl, model);
-            $elt.html(icon_html);
```
I think we should leave the spinner and error icons.

andreasn replied:
I agree.

mvollmer commented:
I like the blue (i) icon since it tells me that there is more to reveal here.  Otherwise tooltips are like easter eggs: I have to discover them and remember where they are.

martinpitt commented:
So we could keep the blue/red (i) vs. spinner icon and attach the tooltip there, as now. It would again be a bit of a special notification case, but even with that it would be much more consistent than what we have now. @garrett, @andreasn, any opinion?

andreasn commented:
I don't have a super-strong stance on (i)'s vs. hover tooltips.
The Patternfly guidelines for info-popups doesn't have a lot of details.
I agree that we need to keep the spinner and error states though.

mvollmer commented:
If you zoom with the browser (Ctrl-plus), the machine id tooltip looks like this:

![tooltip-zoomed](https://user-images.githubusercontent.com/3228183/37586279-90168b18-2b64-11e8-877a-6de17a5518f6.png)

Note that the machine id is truncated inside the tooltip.

The one pixel gap between triangle and actuall tooltip is probably caused by rounding errors somewhere.  I have this in `storage.css`, it probably is time to promote it to `lib/..`.

```
/* Layout fix for tooltip arrows.
 *
 * Sometimes when there are non-integer coordinates involved
 * somewhere, rounding seems to make it so that there is a visible one
 * pixel gap between tooltip-inner and tooltip-arrow.  We avoid that
 * gap by letting the arrow and inner parts overlap by one pixel.
 *
 * This happens within the caption of the content listing.
 */

.tooltip.top .tooltip-arrow {
    bottom: 1px;
}

.tooltip.bottom .tooltip-arrow {
    top: 1px;
}
```

mvollmer commented:
Clicking on the "Hardware" link and then coming back leaves the tooltip open.  It closes when the mouse touches the link again.

mvollmer commented on pkg/systemd/host.css:
```diff
@@ -212,6 +212,11 @@ a.disabled {
 
+#system_information_machine_id_row .tooltip-inner {
+    font-family: monospace;
+    max-width: 35ex;
```
What is the intention of this?  Can't we just let the tooltip be as large as the browser/bootstrap makes it?  Also, `ex` is for height, and `em` is for width, no?

martinpitt replied:
The intention was to make this wide enough to fit the 33 characters of a machine id. Without this, tooltips have a too narrow max width, so that the machine ID doesn't even fit on default zoom.

Indeed I should  use `em` instead of `ex`, this fixes the truncation. Thanks for spotting!

mvollmer replied:
> Without this, tooltips have a too narrow max width

Ah, so this overrides an existing style, makes sense.

mvollmer commented on pkg/realmd/operation.js:
```diff
@@ -413,7 +413,9 @@
-        var element = $("<a>");
+        var element = $("<span>");
+        var link = $("<a>");
+        element.append(link);
```
`update_realm_privileged` still puts `disabled` on element, not on the link.

martinpitt replied:
`update_privileged()` itself needs to be modified for that, but fortunately all other usages of this involve buttons (where notifications work on disabled buttons). Fixed locally.

martinpitt commented:
> Clicking on the "Hardware" link and then coming back leaves the tooltip open. It closes when the mouse touches the link again.

Fixed.

> Note that the machine id is truncated inside the tooltip.

I moved the css fix to pkg/lib/tooltip.css in a new commit. There it's automaticaly being used for the `Tooltip` react component (thus on the storage, packagekit, and other pages). I then adjusted the existing commit to slurp in tooltip.css.

This should cover all your findings, thanks @mvollmer! Can you PTAL?

mvollmer commented:
> fixed all of the issues you found, thanks!

Thanks!

martinpitt commented:
> Why is the time sync notification below?

I don't know, I'm afraid. I kept the existing positioning. Note that having it on the side might not work well with the single-column layout of mobile mode/narrow windows?

garrett commented:
Yeah, that's probably true about mobile mode.

What about on top instead of below? With the tooltip below, it makes it difficult to scan downward while using the mouse. (It's basically covering up the next item.)

garrett commented:
Oops. Trying to retrigger the review thing in GitHub is a bit odd.
